tests/components/tasmota/test_device_trigger.py

```diff
@@ -0,0 +1,832 @@
+    assert len(calls) == 1
+    assert calls[0].data["some"] == "short_press"
+
+    # Fake long press.
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/stat/SWITCH3T", '{"TRIG":"HOLD"}')
+    await hass.async_block_till_done()
+    assert len(calls) == 2
+    assert calls[1].data["some"] == "long_press"
+
+
+async def test_if_fires_on_mqtt_message_late_discover(
+    hass, device_reg, calls, mqtt_mock, setup_tasmota
+):
+    """Test triggers firing of MQTT device triggers discovered after setup."""
+    # Discover a device without device triggers
+    config1 = copy.deepcopy(DEFAULT_CONFIG)
+    config1["swc"][0] = -1
+    mac = config1["mac"]
+
+    # Discover a device with 2 device triggers
+    config2 = copy.deepcopy(DEFAULT_CONFIG)
+    config2["swc"][0] = 0
+    config2["swc"][3] = 9
+
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{mac}/config", json.dumps(config1))
+    await hass.async_block_till_done()
+
+    device_entry = device_reg.async_get_device(set(), {("mac", mac)})
+
+    assert await async_setup_component(
+        hass,
+        automation.DOMAIN,
+        {
+            automation.DOMAIN: [
+                {
+                    "trigger": {
+                        "platform": "device",
+                        "domain": DOMAIN,
+                        "device_id": device_entry.id,
+                        "discovery_id": "00000049A3BC_switch_1_TOGGLE",
+                        "type": "button_short_press",
+                        "subtype": "switch_1",
+                    },
+                    "action": {
+                        "service": "test.automation",
+                        "data_template": {"some": ("short_press")},
+                    },
+                },
+                {
+                    "trigger": {
+                        "platform": "device",
+                        "domain": DOMAIN,
+                        "device_id": device_entry.id,
+                        "discovery_id": "00000049A3BC_switch_4_HOLD",
+                        "type": "switch_4",
+                        "subtype": "button_double_press",
+                    },
+                    "action": {
+                        "service": "test.automation",
+                        "data_template": {"some": ("double_press")},
+                    },
+                },
+            ]
+        },
+    )
+
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{mac}/config", json.dumps(config2))
+    await hass.async_block_till_done()
+
+    # Fake short press.
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/stat/SWITCH1T", '{"TRIG":"TOGGLE"}')
+    await hass.async_block_till_done()
+    assert len(calls) == 1
+    assert calls[0].data["some"] == "short_press"
+
+    # Fake long press.
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/stat/SWITCH4T", '{"TRIG":"HOLD"}')
+    await hass.async_block_till_done()
+    assert len(calls) == 2
+    assert calls[1].data["some"] == "double_press"
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6333224)